optimum/fx/parallelization/utils.py

```diff
@@ -0,0 +1,211 @@
+import operator
+import importlib
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from typing import Dict, Callable, List, Union
+from torch.fx import Node, Graph
+from functools import wraps
+from collections import defaultdict
+from itertools import chain
+from .core import (
+    ParallelExecutionCtx,
+    ParameterMapping,
+    ParameterMeta,
+)
+
+def is_linear(node: Node) -> bool:
+    if node.op != 'call_module':
+        return False
+    mod = node.graph.owning_module
+    return isinstance(mod.get_submodule(node.target), nn.Linear)
+
+def is_shape_consumer(node: Node) -> bool:
+    if node.op == 'call_method':
+        return node.target in {'view', 'reshape', 'expand', 'resize', 'resize_'}
+    elif node.op == 'call_function':
+        return node.target in {torch.reshape}
+
+def is_transpose(node: Node) -> bool:
+    if node.op == 'call_method':
+        return node.target in {'transpose', 'transpose_'}
+    elif node.op == 'call_function':
+        return node.target is torch.transpose
+    return False
+
+def is_permute(node: Node) -> bool:
+    if node.op == 'call_method':
+        return node.target in {'permute'}
+    elif node.op == 'call_function':
+        return node.target is torch.permute
+    return False
+
+def is_getitem(node: Node) -> bool:
+    return node.op == 'call_function' and node.target is operator.getitem
+
+def is_output(node: Node) -> bool:
+    return node.op == 'output'
+
+def is_shape_generator(node: Node) -> bool:
+    return node.op == 'call_method' and node.target == 'size'
```

Comment:
But `tensor.shape` would not get caught here no?

Reply:
`.shape` will be traced as `size`